File: filly-core/src/widgets/file_picker.rs
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, ListItem, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;
use std::fs;
use std::path::{Path, PathBuf};

pub struct FilePickerWidget {
    pub title: String,
    pub current_dir: PathBuf,
    pub entries: Vec<PathBuf>,
    pub selected: usize,
    pub filter_ext: String,
    pub mode: PickerMode,
    dirty: bool,
}

#[derive(PartialEq)]
pub enum PickerMode {
    Browsing,
    ConfirmQuit,
}

impl FilePickerWidget {
    pub fn new(title: String, start_dir: Option<String>, filter: Option<String>) -> Self {
        let current_dir = start_dir
            .map(PathBuf::from)
            .unwrap_or_else(|| PathBuf::from("/"));
        let filter_ext = filter.unwrap_or_default();
        let entries = get_entries(&current_dir, &filter_ext);
        Self {
            title,
            current_dir,
            entries,
            selected: 0,
            filter_ext,
            mode: PickerMode::Browsing,
            dirty: true,
        }
    }

    fn navigate_to(&mut self, dir: PathBuf) {
        self.current_dir = dir;
        self.entries = get_entries(&self.current_dir, &self.filter_ext);
        self.selected = 0;
        self.dirty = true;
    }
}

impl Widget for FilePickerWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = ((area.width as f32 * 0.7) as u16).min(area.width.saturating_sub(2));
        let box_h = ((area.height as f32 * 0.75) as u16).min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;

        let mut children = Vec::new();
        children.push(RenderNode::Text {
            rect: Rect::new(1, 0, box_w - 2, 1),
            content: format!("{}: {}", self.title, self.current_dir.display()),
            alignment: Alignment::Center,
            role: None, label: None, description: None, style: TextStyle {
                bold: true,
                fg: Some("green".into()),
                ..TextStyle::normal()
            },
        });

        let list_y = 1;
        let list_h = box_h.saturating_sub(list_y).saturating_sub(2);
        let items: Vec<ListItem> = self
            .entries
            .iter()
            .map(|entry| {
                let is_dir = entry.is_dir();
                let prefix = if is_dir { "[DIR] " } else { "      " };
                let name = entry
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .to_string();
                ListItem {
                    label: format!("{}{}", prefix, name),
                    meta: None,
                }
            })
            .collect();

        children.push(RenderNode::List {
            rect: Rect::new(1, list_y, box_w - 2, list_h),
            items,
            selected: Some(self.selected),
            highlight_style: TextStyle::selected(), role: None, label: None, description: None,
        });

        let help_y = box_h.saturating_sub(1);
        children.push(RenderNode::Text {
            rect: Rect::new(1, help_y, box_w - 2, 1),
            content: "Up/Down:move  Enter:open/select  Backspace:parent  Esc:cancel".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(), role: None, label: None, description: None,
            children,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        match (&self.mode, event) {
            (PickerMode::ConfirmQuit, Event::Key(code, _)) => match code {
                KeyCode::Char('y') | KeyCode::Char('Y') => {
                    EventResult::Response(WidgetResponse {
                        result: None,
                        cancelled: true,
                        error: None,
                    })
                }
                _ => {
                    self.mode = PickerMode::Browsing;
                    self.dirty = true;
                    EventResult::Handled
                }
            },
            (PickerMode::Browsing, Event::Key(code, _)) => match code {
                KeyCode::Esc => {
                    self.mode = PickerMode::ConfirmQuit;
                    self.dirty = true;
                    EventResult::Handled
                }
                KeyCode::Up => {
                    if self.selected > 0 {
                        self.selected -= 1;
                        self.dirty = true;
                    }
                    EventResult::Handled
                }
                KeyCode::Down => {
                    if self.selected + 1 < self.entries.len() {
                        self.selected += 1;
                        self.dirty = true;
                    }
                    EventResult::Handled
                }
                KeyCode::Enter => {
                    if let Some(entry) = self.entries.get(self.selected) {
                        if entry.is_dir() {
                            self.navigate_to(entry.clone());
                        } else {
                            return EventResult::Response(WidgetResponse {
                                result: Some(serde_json::Value::String(
                                    entry.to_string_lossy().to_string(),
                                )),
                                cancelled: false,
                                error: None,
                            });
                        }
                    }
                    EventResult::Handled
                }
                KeyCode::Backspace => {
                    if let Some(parent) = self.current_dir.parent() {
                        self.navigate_to(parent.to_path_buf());
                    }
                    EventResult::Handled
                }
                _ => EventResult::Unhandled,
            },
            _ => EventResult::Unhandled,
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

fn get_entries(dir: &Path, filter: &str) -> Vec<PathBuf> {
    let mut entries: Vec<PathBuf> = if let Ok(read_dir) = fs::read_dir(dir) {
        read_dir
            .filter_map(|e| e.ok().map(|e| e.path()))
            .filter(|p| {
                if filter.is_empty() {
                    true
                } else if p.is_dir() {
                    true
                } else {
                    p.extension().map(|ext| ext == filter).unwrap_or(false)
                }
            })
            .collect()
    } else {
        vec![]
    };
    entries.sort_by(|a, b| {
        let a_dir = a.is_dir();
        let b_dir = b.is_dir();
        if a_dir && !b_dir {
            std::cmp::Ordering::Less
        } else if !a_dir && b_dir {
            std::cmp::Ordering::Greater
        } else {
            a.file_name().cmp(&b.file_name())
        }
    });
    entries
}

File: filly-core/widgets/file_picker.c
#include <stdio.h>
#include "file_picker.h"
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>

typedef struct {
    char *title;
    char *current_dir;
    char *filter_ext;
    char **entries;
    int entry_count;
    int selected;
    bool *is_dir;
    bool dirty;
} FilePickerData;

static void fp_populate(FilePickerData *d) {
    for (int i = 0; i < d->entry_count; i++) free(d->entries[i]);
    free(d->entries); free(d->is_dir);
    d->entries = NULL; d->is_dir = NULL; d->entry_count = 0;

    DIR *dir = opendir(d->current_dir);
    if (!dir) return;
    struct dirent *entry;
    int cap = 64;
    d->entries = malloc(cap * sizeof(char *));
    d->is_dir = malloc(cap * sizeof(bool));
    while ((entry = readdir(dir))) {
        if (strcmp(entry->d_name, ".") == 0) continue;
        char path[4096];
        snprintf(path, sizeof(path), "%s/%s", d->current_dir, entry->d_name);
        struct stat st;
        if (stat(path, &st) != 0) continue;
        bool isdir = S_ISDIR(st.st_mode);
        if (!isdir && d->filter_ext && strlen(d->filter_ext)) {
            const char *ext = strrchr(entry->d_name, '.');
            if (!ext || strcmp(ext + 1, d->filter_ext) != 0) continue;
        }
        if (d->entry_count >= cap) { cap *= 2; d->entries = realloc(d->entries, cap * sizeof(char *)); d->is_dir = realloc(d->is_dir, cap * sizeof(bool)); }
        d->entries[d->entry_count] = strdup(entry->d_name);
        d->is_dir[d->entry_count] = isdir;
        d->entry_count++;
    }
    closedir(dir);
    if (d->selected >= d->entry_count) d->selected = d->entry_count > 0 ? d->entry_count - 1 : 0;
}

static void file_picker_render(Widget *self, Rect area, RenderTree *out) {
    FilePickerData *d = (FilePickerData *)(self + 1);
    memset(out, 0, sizeof(*out));
    int box_w = (int)(area.w * 0.7f);
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_h = (int)(area.h * 0.75f);
    if (box_h > area.h - 2) box_h = area.h - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    out->accessible.role = strdup("dialog");
    out->accessible.label = strdup(d->title ? d->title : "File Picker");

    RenderTree *children = calloc(3 + 1, sizeof(RenderTree));
    children[0].type = RNODE_TEXT;
    children[0].rect = rect_new(1, 0, box_w - 2, 1);
    char title_buf[512];
    snprintf(title_buf, sizeof(title_buf), "%s: %s", d->title, d->current_dir);
    children[0].text.content = strdup(title_buf);
    children[0].text.align = ALIGN_CENTER;
    children[0].text.style = textstyle_selected();
    children[0].accessible.role = strdup("heading");
    children[0].accessible.label = strdup(title_buf);

    children[1].type = RNODE_LIST;
    children[1].rect = rect_new(1, 1, box_w - 2, box_h - 3);
    children[1].list.item_count = d->entry_count;
    children[1].list.items = malloc(d->entry_count * sizeof(ListItem));
    for (int i = 0; i < d->entry_count; i++) {
        char label[512];
        snprintf(label, sizeof(label), "%s %s", d->is_dir[i] ? "[DIR]" : "     ", d->entries[i]);
        children[1].list.items[i] = listitem_new(label);
    }
    children[1].list.selected = d->selected;
    children[1].list.highlight = textstyle_selected();
    children[1].accessible.role = strdup("listbox");

    children[2].type = RNODE_TEXT;
    children[2].rect = rect_new(1, box_h - 2, box_w - 2, 1);
    children[2].text.content = strdup("Up/Down:move  Enter:open/select  Backspace:parent  Esc:cancel");
    children[2].text.align = ALIGN_CENTER;
    children[2].text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = 3;
}

static EventResult file_picker_handle_event(Widget *self, Event *ev, Backend *backend) {
    (void)backend;
    FilePickerData *d = (FilePickerData *)(self + 1);
    if (ev->type != EVENT_KEY) return event_result_unhandled();
    switch (ev->code) {
        case KEY_ESC:
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        case KEY_UP:
            if (d->selected > 0) { d->selected--; d->dirty = true; }
            return event_result_handled();
        case KEY_DOWN:
            if (d->selected + 1 < d->entry_count) { d->selected++; d->dirty = true; }
            return event_result_handled();
        case KEY_ENTER:
            if (d->selected < d->entry_count) {
                if (d->is_dir[d->selected]) {
                    char *new_path = malloc(strlen(d->current_dir) + strlen(d->entries[d->selected]) + 2);
                    sprintf(new_path, "%s/%s", d->current_dir, d->entries[d->selected]);
                    free(d->current_dir);
                    d->current_dir = new_path;
                    fp_populate(d);
                    d->dirty = true;
                } else {
                    char *full = malloc(strlen(d->current_dir) + strlen(d->entries[d->selected]) + 2);
                    sprintf(full, "%s/%s", d->current_dir, d->entries[d->selected]);
                    return event_result_response((WidgetResponse){ .result = cJSON_CreateString(full), .cancelled = false, .error = NULL });
                }
            }
            return event_result_handled();
        case KEY_BACKSPACE: {
            char *slash = strrchr(d->current_dir, '/');
            if (slash && slash != d->current_dir) {
                *slash = '\0';
                fp_populate(d);
                d->dirty = true;
            }
            return event_result_handled();
        }
        default: return event_result_unhandled();
    }
}

static bool fp_is_dirty(Widget *self) { return ((FilePickerData *)(self + 1))->dirty; }
static void fp_clear_dirty(Widget *self) { ((FilePickerData *)(self + 1))->dirty = false; }
static void fp_destroy(Widget *self) {
    FilePickerData *d = (FilePickerData *)(self + 1);
    free(d->title); free(d->current_dir); free(d->filter_ext);
    for (int i = 0; i < d->entry_count; i++) free(d->entries[i]);
    free(d->entries); free(d->is_dir);
}

Widget *file_picker_widget_new(const char *title, const char *start_dir, const char *filter_ext) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(FilePickerData));
    w->vtable.render = file_picker_render;
    w->vtable.handle_event = file_picker_handle_event;
    w->vtable.is_dirty = fp_is_dirty;
    w->vtable.clear_dirty = fp_clear_dirty;
    w->vtable.destroy = fp_destroy;
    FilePickerData *d = (FilePickerData *)(w + 1);
    d->title = strdup(title);
    d->current_dir = start_dir ? strdup(start_dir) : strdup("/");
    d->filter_ext = filter_ext ? strdup(filter_ext) : NULL;
    d->entries = NULL; d->entry_count = 0; d->is_dir = NULL; d->selected = 0;
    fp_populate(d);
    d->dirty = true;
    return w;
}